### ISSUE TITLE
Repeater should be added before any other content (or allow it)

### ISSUE THREAD

**Issue by estani** [prawnpdf/prawn#1135]
We use different background images for different pages, adding it with a repeater this way:
```
repeat(->(pg) { pg >= 2 }) do
      blend_mode(:Overlay) do
        load_background_image
      end
    end
```
We need to do it in Overlay mode, because the text is rendered behind the image. The overlay affects the rendering so we want to get rid of it.

I think what is happening is that the repeat block (or stamp) is called after the page has been rendered, not before.

**Comment by pointlessone** [comment]
Repeaters are ran right before the document gets rendered. That is, after all content has been placed. That is by design.

Is there a reason you can not render images before placing the content?

**Comment by estani** [comment]
You mean outside of a repeater? (because as you said, repeaters are called "after all content is placed", so rendering images "before" means not using a repeater, right?)
That will require some callback, after a new page gets rendered. Pretty much like a repeater... hence that's what the PR is about.
Is there another way to get an image pasted depending on the page number "before" the content is placed that I'm not aware of? (perhaps that was your question? In that case the reason I can't do it, is because I don't know how).

To summarize the problem again: A repeater is like a stamp (it is implemented like it). Now you might want to have a stamp behind some text or above, the later is what's implemented, for the former there's the PR.

I'll be glad to make changes if that's required. I did try to follow the code logic as good as I could, but I'm by no means a prawn expert...

**Comment by pointlessone** [comment]
There are a number of solutions.

1. You can use document background:
   ```ruby
   Prawn::Document.generate "test.pdf", background: 'background.png' do
     text 'foo'

     start_new_page
     text 'bar'

     start_new_page
     text 'baz'
   end
   ```
2. You can use `on_page_create` callback:
   ```ruby
   Prawn::Document.generate "test.pdf" do
     on_page_create do
       image 'data/images/prawn.png'
     end

     text 'foo'

     start_new_page
     text 'bar'

     start_new_page
     text 'baz'
   end
   ```

   However note that callback is not called for the first page of the document.

**Comment by estani** [comment]
If I understood your code and comment correctly, this would not work:
1) cannot use different backgrounds for different pages
2) cannot add a background to the first page (from your comment)

From your comments the best option would be to call on_page_create for the first page too (and by the way pass some context information, like the page count)

**Comment by pointlessone** [comment]
It's not quite possible to call `on_page_create` because the first page is created before it can be defined. As a work around you can pain the background on the first page manually.

As for the page number, it's accessible via `page_number` method.

Here's an example:
```ruby
Prawn::Document.generate "test.pdf" do
  # This is a proc that draws background
  page_bg = proc do
    image 'background.png'
    text_box page_number.to_s # Show the page number
  end

  on_page_create(&page_bg) # Set the proc as callback

  # Manually call the background drawing proc on the first page
  # callback is called in a `float` to not disturb the flow of the page, here we do it for the same reason
  float { page_bg.call }
  move_down 20 
  text 'foo'

  start_new_page
  move_down 20
  text 'bar'

  start_new_page
  move_down 20
  text 'baz'
end
```

**Comment by buncis** [comment]
I have similar problem

I have an image that has a header and background for the page
```
  repeat(:all) do
    on_page_create do
      image "app/assets/images/header_pdf.png", :width => bounds.width
    end
  end
```
the problem with this is the image will be put out as foreground instead background


the workaround are using `on_page_create` problem with `on_page_create` is its not executed on first page

so you will need to add it manualy

so the complete solution are

```
    image "app/assets/images/header_pdf.png", :width => bounds.width
    repeat(:all) do 
      on_page_create do
        image "app/assets/images/header_pdf.png", :width => bounds.width
      end
    end
```

**Comment by pointlessone** [comment]
I don't think placing `on_page_create` into a repeater achieves anything useful.

I agree that this is a confusing behaviour but it's a nature of Prawn. You can think of Prawn as being a streaming library. Document is constructed as the code executes. The fact that you can jump around pages doesn't change this fact.

Streaming nature dictates how the document is created and in which order it is being constructed. For historical reasons (mostly) Prawn creates first page on initialization. That is before any `on_page_create` callback could be defined. 

One work around might be to pass `skip_page_creation: true` option to Prawn document constructor. That will still create a page but it won't be added to the document sequence. You'll have an opportunity to define the callback but you'll have to manually add the first page.

```ruby
require 'open-uri'

Prawn::Document.generate("test.pdf", skip_page_creation: true) do
  on_page_create do
    image URI.open("https://prawnpdf.org/images/prawn.png")
  end

  1.upto(3) do |n|
    start_new_page
    text "Page #{n}\n" * 10
  end
end
```

This produces a document with 3 pages each having an image in the background and the text in the foreground.

[test.pdf](https://github.com/prawnpdf/prawn/files/7555074/test.pdf)

**Comment by buncis** [comment]
yeah I have refine my code the only problem that exist to me is 

how to make the text started at certain cursor (below the repeated image ideally)?

I could use `move_cursor_to` if I use `start_new_page` manually 

problem is when the page creation is created by flowing content the cursor is not moved at all in that case

**Comment by pointlessone** [comment]
You can try setting margins to leave enough space for your header image.

**Comment by buncis** [comment]
the only way to setup margin is only by giving parameter to `Prawn::Document.generate()` and `start_new_page()` right?

**Comment by pointlessone** [comment]
> the only way to setup margin is only by giving parameter to Prawn::Document.generate() and start_new_page() right?

You can also use `bounding_box` with a block or assign `bounds` on the document.

**Comment by buncis** [comment]
thank you so much for helping and maintaining prawn
